### PR TITLE
3.4 bloom continuous rebuild

### PR DESCRIPTION
**Pull request by ragadeeshu**

### CODE REVIEW

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 1]
```diff
@@ -92,36 +94,47 @@ public FulltextProviderImpl( GraphDatabaseService db, Log log, AvailabilityGuard
...
+            while ( !isAvailable && !availabilityGuard.isShutdown() );
+            while ( flipper )
+            {
+                try
+                {
+                    flip();
+                }
+                catch ( IOException e )
+                {
+                    e.printStackTrace();
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: You could also write it to the `log` instead.

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 2]
```diff
@@ -144,19 +157,22 @@ public void awaitPopulation()
...
-        LuceneFulltext index = factory.createFulltextIndex( identifier, type, properties );
-        register( index );
+        LuceneFulltext a = factory.createFulltextIndex( identifier, ASIDE, type, properties );
+        register( a, ASIDE );
+        LuceneFulltext b = factory.createFulltextIndex( identifier, BSIDE, type, properties );
+        register( b, BSIDE );
     }
 
-    private void register( LuceneFulltext fulltextIndex ) throws IOException
+    private void register( LuceneFulltext fulltextIndex, int side ) throws IOException
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: Is there a reason you don't pass both sides to one register call, so both can be assigned atomically under the `configurationLock`?

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 3]
```diff
@@ -337,25 +324,100 @@ public void registerFileListing( NeoStoreFileListing fileListing )
...
+                index.flush();
+                ResourceIterator<File> snapshot = index.snapshot();
+                snapshots.add( snapshot );
+                files.addAll( getSnapshotFilesMetadata( snapshot ) );
+            }
+        }
+        finally
+        {
+            lock.unlock();
         }
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: Don't we need to keep the configuration read lock for the duration of the snapshotting? What would otherwise happen if we drop the index while reading the snapshot?

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 4]
```diff
@@ -367,7 +429,15 @@ private Resource snapshotStoreFiles( Collection<StoreFileMetadata> files ) throw
...
     {
-        db.unregisterTransactionEventHandler( fulltextTransactionEventUpdater );
+        flipper = false;
+        try
+        {
+            flipperJob.waitTermination();
+        }
+        catch ( InterruptedException | ExecutionException e )
+        {
+            e.printStackTrace();
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: This should go to the log as well.

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 5]
```diff
@@ -367,7 +429,15 @@ private Resource snapshotStoreFiles( Collection<StoreFileMetadata> files ) throw
     @Override
     public void close()
     {
-        db.unregisterTransactionEventHandler( fulltextTransactionEventUpdater );
+        flipper = false;
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: I would call this `closed` and invert the values used.

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 6]
```diff
@@ -337,25 +324,100 @@ public void registerFileListing( NeoStoreFileListing fileListing )
...
+            {
+                // Save the last committed transaction, then drain the update queue to make sure that we have applied _at least_ the commits the config claims.
+                index.saveConfiguration( transactionIdStore.getLastCommittedTransactionId() );
+                try
+                {
+                    applier.writeBarrier().awaitCompletion();
+                }
+                catch ( ExecutionException e )
+                {
+                    throw new IOException( "Unable to prepare index for snapshot." );
```
Comment (chrisvest) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java]: Please include the cause.

**Review by chrisvest — COMMENTED** [review]
These are the comments I have for now. It's hard to follow what's going on.

**Comment by chrisvest** [comment]
The PageCacheWarmer initially also used a/b files, but ran into loads of problems with backup and store copy. I would prefer if you adopted the same incrementing-sequences-of-files plus reference counts for the file listing solution that PCW ended up with.

**Review by chrisvest — APPROVED** [review]

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 7]
```diff
@@ -337,37 +336,174 @@ public void registerFileListing( NeoStoreFileListing fileListing )
...
+                for ( WritableFulltext writableFulltext : writableRelationshipIndices[tentativeSide].values() )
+                {
+                    writableFulltext.drop();
+                    writableFulltext.open();
+                    populations.add( applier.populateRelationships( writableFulltext, db ) );
+                }
+            }
+            finally
+            {
+                configurationLock.writeLock().unlock();
```
Comment (klaren) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java:390]: There are a couple of lines not covered by this `finally`, I recommend to move the `try` up to include everything.

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 8]
```diff
@@ -337,37 +336,174 @@ public void registerFileListing( NeoStoreFileListing fileListing )
...
     }
 
-    private Resource snapshotStoreFiles( Collection<StoreFileMetadata> files ) throws IOException
+    @Override
+    public void awaitFlip()
     {
-        final Collection<ResourceIterator<File>> snapshots = new ArrayList<>();
-        for ( WritableFulltext index : Iterables.concat( writableNodeIndices.values(), writableRelationshipIndices.values() ) )
+        int startSide = tentativeSide;
+        synchronized ( this )
```
Comment (klaren) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java:343]: You can synchronize the method instead.

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 9]
```diff
@@ -61,18 +60,23 @@
  */
 public class FulltextProviderImpl implements FulltextProvider
 {
+    private static final int ASIDE = 0;
```
Comment (klaren) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java:63]: I read this as "aside", what do you think about having `A_SIDE` instead? :wink:

File: enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java [hunk 10]
```diff
@@ -337,37 +336,174 @@ public void registerFileListing( NeoStoreFileListing fileListing )
...
+        configurationLock.writeLock().lock();
+        try
         {
-            // Save the last committed transaction, then drain the update queue to make sure that we have applied _at least_ the commits the config claims.
-            index.saveConfiguration( transactionIdStore.getLastCommittedTransactionId() );
+            List<AsyncFulltextIndexOperation> populations = new ArrayList<>();
+            synchronized ( this )
+            {
+                int currentSide = side;
+                tentativeSide = (currentSide + 1) % 2;
```
Comment (klaren) [inline enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java:369]: No need to have the `currentSide` here, also, this looks strange since we have `A_SIDE` and `B_SIDE`. How about
```
tentativeSide = side == A_SIDE ? B_SIDE : A_SIDE;
```

**Review by klaren — APPROVED** [review]
A few cleanup comments, feel free to ignore.
